### PR TITLE
Add new cryptoTransfer selector 

### LINKED ISSUE #4086: Create new selector and decoder for new cryptoTransfer
### Problem

Epic #4084 

Add support for new selector for the cryptoTransfer function as the signature has changed with the addition of the isApproval boolean parameter and the transferList for hbar transfers

### Solution

Add support for the new selector for cryptoTransfer.  
Create a new decoder that decodes the new function.

Merge hbar transfers into the transactionBody.

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by lukelee-sl**
Signed-off-by: lukelee-sl <luke.lee@swirldslabs.com>

**Description**:
The new signature of the cryptoTransfer function means that a new selector must be support.
Leave the existing selector for backwards compatibility.

After the parameters are decoded, update the transaction body as necessary.

Add a property to feature gate hbar transfers

**Related issue(s)**:

Fixes #4086 

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-node/hedera-mono-service/src/test/java/com/hedera/services/store/contracts/precompile/TransferPrecompilesTest.java [hunk 1]
```diff
@@ -1554,20 +1566,144 @@ void decodeTransferNFTsInput() {
         assertEquals(0, hbarTransfers.size());
     }
 
-    private void givenFrameContext() {
```
Comment (lukelee-sl) [inline hedera-node/hedera-mono-service/src/test/java/com/hedera/services/store/contracts/precompile/TransferPrecompilesTest.java:1557]: this method was removed as it was no longer referenced.

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java [hunk 2]
```diff
@@ -310,30 +341,67 @@ public static CryptoTransferWrapper decodeCryptoTransfer(
...
+
+            final var abiAdjustments = (Tuple[]) tupleNested.get(1);
+            if (abiAdjustments.length > 0) {
+                fungibleTransfers =
+                        bindFungibleTransfersFrom(tokenType, abiAdjustments, aliasResolver);
+            }
+            final var abiNftExchanges = (Tuple[]) tupleNested.get(2);
+            if (abiNftExchanges.length > 0) {
+                nftExchanges = bindNftExchangesFrom(tokenType, abiNftExchanges, aliasResolver);
+            }
```
Comment (dimitar-dinev) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java:399]: Shouldn't we also alter [`bindFungibleTransferFrom` and `bindNftExchangesFrom` ](https://github.com/hashgraph/hedera-services/blob/53d7b5249d54f7878a452872a0f09da434d8b9b4/hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/codec/DecodingFacade.java#L175-L203)to take into account [the new `isApproval` field in the Solidity structs](https://github.com/hashgraph/hedera-smart-contracts/blob/7374d6ead7e5c13cd726e553d269507043241f77/contracts/hts-precompile/IHederaTokenService.sol#L46) or is that planned for a future PR?

Comment (lukelee-sl) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java:399]: This planned for a future pr but thanks for pointing this out.

File: hedera-node/hedera-mono-service/src/main/resources/bootstrap.properties [hunk 3]
```diff
@@ -83,6 +83,7 @@ contracts.scheduleThrottleMaxGasLimit=5000000
 contracts.sidecars=CONTRACT_STATE_CHANGE,CONTRACT_BYTECODE
 contract.storageSlotPriceTiers=0til100M,2000til450M
 contracts.throttle.throttleByGas=true
+contracts.precompile.crypto.transfer.has.hbar=false
```
[the remark this answers is not in the document]
Comment (lukelee-sl) [reply hedera-node/hedera-mono-service/src/main/resources/bootstrap.properties:86]: That's a good suggestion.

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java [hunk 4]
```diff
@@ -161,6 +176,11 @@ public TransactionBody.Builder body(
 
         transactionBody =
                 syntheticTxnFactory.createCryptoTransfer(transferOp.tokenTransferWrappers());
+        if (hasHbarTransfer) {
+            transactionBody.mergeFrom(
+                    syntheticTxnFactory.createCryptoTransferForHbar(transferOp.transferWrapper()));
+        }
+
```
[the remark this answers is not in the document]
Comment (lukelee-sl) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java:183]: Makes sense.  I'll update the code as you suggest.

File: hedera-node/hedera-mono-service/src/main/resources/bootstrap.properties (hunk 3 above)
[the remark this answers is not in the document]
Comment (Nana-EC) [reply hedera-node/hedera-mono-service/src/main/resources/bootstrap.properties]: What about `contracts.precompile.atomicCryptoTransfer.enabled`

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java [hunk 5]
```diff
@@ -310,30 +339,67 @@ public static CryptoTransferWrapper decodeCryptoTransfer(
...
+        List<SyntheticTxnFactory.HbarTransfer> hbarTransfers = new ArrayList<>();
+        final List<TokenTransferWrapper> tokenTransferWrappers = new ArrayList<>();
 
-                tokenTransferWrappers.add(
-                        new TokenTransferWrapper(nftExchanges, fungibleTransfers));
-            }
+        final Tuple[] hbarTransferTuples = ((Tuple) decodedTuples.get(0)).get(0);
+        final var tokenTransferTuples = decodedTuples.get(1);
+
+        if (hbarTransferTuples.length > 0) {
```
Comment (Nana-EC) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java:370]: nit: would be nice to have a `decodeHBarTransfer` to contain the appropriate logic

**Review by Nana-EC — COMMENTED** [review]
LG on first pass, will circle back again

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/codec/DecodingFacade.java [hunk 6]
```diff
@@ -228,6 +241,20 @@ public static void addSignedAdjustment(
...
+            final List<SyntheticTxnFactory.HbarTransfer> hbarTransfers,
+            final AccountID accountID,
+            final long amount,
+            final boolean isApproval) {
+        if (amount > 0) {
+            hbarTransfers.add(
+                    new SyntheticTxnFactory.HbarTransfer(amount, isApproval, null, accountID));
+        } else {
+            hbarTransfers.add(
+                    new SyntheticTxnFactory.HbarTransfer(-amount, isApproval, accountID, null));
```
[the remark this answers is not in the document]
Comment (lukelee-sl) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/codec/DecodingFacade.java:254]: I looked into this as it was bothering me as well.  I'm following the pattern of fungible token transfers which is a descendant of HbarTransfer.   
It turns out that the validation that happens as part of the HAPI transfer code assumes that the amounts are positive and determine which way the tokens/hbars are transferred via see if we are the sender or the receiver.  I believe the validation code throws an exception if the amount is negative.
Later on, when the changes are actually created the amounts for the senders are negated again to correctly have negative amounts.

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/context/properties/PropertyNames.java [hunk 7]
```diff
@@ -119,6 +119,8 @@ private PropertyNames() {
             "contracts.precompile.exportRecordResults";
     public static final String CONTRACTS_PRECOMPILE_HTS_ENABLE_TOKEN_CREATE =
             "contracts.precompile.htsEnableTokenCreate";
+    public static final String CONTRACTS_PRECOMPILE_CRYPTO_TRANSFER2_ENABLED =
```
[the remark this answers is not in the document]
Comment (lukelee-sl) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/services/context/properties/PropertyNames.java:122]: done

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java (hunk 5 above)
Comment (lukelee-sl) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java:370]: done

File: hedera-node/hedera-mono-service/src/main/resources/bootstrap.properties (hunk 3 above)
[the remark this answers is not in the document]
Comment (lukelee-sl) [reply hedera-node/hedera-mono-service/src/main/resources/bootstrap.properties]: done

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (lukelee-sl) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/impl/TransferPrecompile.java]: done

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/codec/DecodingFacade.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (Nana-EC) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/services/store/contracts/precompile/codec/DecodingFacade.java:254]: Gotcha. Thanks for confirming.

**Review by Nana-EC — APPROVED** [review]

**Review by tdermendzhievv — APPROVED** [review]
LG

**Pull request merged**